server/build_event_protocol/build_event_server/build_event_server.go

```diff
@@ -104,20 +104,25 @@ func (s *BuildEventProtocolServer) PublishBuildToolEventStream(stream pepb.Publi
 			defer channel.Close()
 		}
 
-		if err := channel.HandleEvent(in); err != nil {
-			if status.IsAlreadyExistsError(err) {
-				log.Warningf("AlreadyExistsError handling event; this means the invocation already exists and may not be retried: %s", err)
-				return err
+		select {
+		case <-channel.Context().Done():
+			return status.UnavailableError("context cancelled")
```

Comment:
I think in this case the invocations won't be marked as disconnected because you are not using `disconnectWithErr` but you can't use `disconnectWithErr` as it would have the same issue that you are trying to fix.

I think you could use disconnectWithErr if you used the select around the stream recv:
```
var channelDone <-chan struct{}
select {
   case <-channelDone:
      return disconnectWithErr(...)
  case in:=<-streamMsg:
      // handle message
     // set channelDone when channel is created
}

this would allow you to cancel the stream as soon as the shutdown starts instead of only when new messages are received on the stream

Reply:
Done.